lib/msal-common/src/client/OnBehalfOfClient.ts

```diff
@@ -0,0 +1,173 @@
+        const cachedAccessToken = this.readAccessTokenFromCache(request);
+        if (!cachedAccessToken ||
+            TimeUtils.isTokenExpired(cachedAccessToken.expiresOn, this.config.systemOptions.tokenRenewalOffsetSeconds)) {
+            return null;
+        }
+
+        const cachedIdToken = this.readIdTokenFromCache(request);
+        let idTokenObject: IdToken = null; 
+        let cachedAccount: AccountEntity = null;
+        if(cachedIdToken) {
+            idTokenObject = new IdToken(cachedIdToken.secret, this.config.cryptoInterface);
+            const accountKey = AccountEntity.generateAccountCacheKey({
+                homeAccountId: cachedIdToken.homeAccountId,
+                environment: cachedIdToken.environment, 
+                tenantId: cachedIdToken.realm,
+                username: null
+            });
+        
+            cachedAccount = this.cacheManager.getAccount(accountKey);
+        }
+
+        return ResponseHandler.generateAuthenticationResult({
+            account: cachedAccount,
+            accessToken: cachedAccessToken,
+            idToken: cachedIdToken,
+            refreshToken: null
+        }, idTokenObject, true);
+    }
+
+    private readAccessTokenFromCache(request: OnBehalfOfRequest): AccessTokenEntity {
+        const accessTokenFilter: CredentialFilter = {
+            environment: this.authority.canonicalAuthorityUrlComponents.HostNameAndPort,
+            credentialType: CredentialType.ACCESS_TOKEN,
+            clientId: this.config.authOptions.clientId,
+            realm: this.authority.tenant,
+            target: this.scopeSet.printScopesLowerCase(),
+            oboAssertion: request.oboAssertion
+        };
+
+        const credentialCache: CredentialCache = this.cacheManager.getCredentialsFilteredBy(accessTokenFilter);
+        const accessTokens = Object.keys(credentialCache.accessTokens).map(key => credentialCache.accessTokens[key]);
+        if (accessTokens.length < 1) {
+            return null;
+        }
+        return accessTokens[0] as AccessTokenEntity;
+    }
+    
+    private readIdTokenFromCache(request: OnBehalfOfRequest): IdTokenEntity {
+        const accessTokenFilter: CredentialFilter = {
+            environment: this.authority.canonicalAuthorityUrlComponents.HostNameAndPort,
+            credentialType: CredentialType.ID_TOKEN,
+            clientId: this.config.authOptions.clientId,
+            realm: this.authority.tenant,
+            oboAssertion: request.oboAssertion
+        };
+
+        const credentialCache: CredentialCache = this.cacheManager.getCredentialsFilteredBy(accessTokenFilter);
+        const idTokens = Object.keys(credentialCache.idTokens).map(key => credentialCache.idTokens[key]);
+        // When acquiring a token on behalf of an application, there might not be an id token in the cache
+        if (idTokens.length < 1) {
+            return null;
+        }
+        return idTokens[0] as AccessTokenEntity;
```

Comment:
Is this a typo? This should be `IdTokenEntity`.

Reply:
Yep, this should be IdTokenEntity. I've updated. Same as above.